File: solutions/ruby/darts/2/darts.rb
class Darts

  def self.score x, y
    new(x, y).score
  end

  private

  attr_accessor :x, :y

  def initialize x, y
    self.x, self.y = x, y
  end

  def distance_from_center
    Math.hypot x, y
  end

  def calculate_impact_region
    distance = distance_from_center
    return :miss          if distance > 10
    return :bullseye      if distance <= 1
    return :inner_circle  if distance <= 5
    :middle_circle
  end

  public

  def score
    case calculate_impact_region
    when :bullseye      then 10
    when :inner_circle  then 5
    when :middle_circle then 1
    else 0
    end
  end

end

File: solutions/c/darts/2/darts.c
#include "darts.h"
#include <math.h>

static const uint8_t inner_circle  = 10;
static const uint8_t middle_circle = 5;
static const uint8_t outer_circle  = 1;
static const uint8_t missed        = 0;

static inline float distance_from_center(coordinate_t impact) {
    return impact.x * impact.x + impact.y * impact.y;
}

uint8_t score(coordinate_t impact) {
    // Comparing distance² against radii²,
    // (1²=1, 5²=25, 10²=100) avoids a sqrtf call.
    float distance = distance_from_center(impact);
    if      (distance <= 1.0f)   return inner_circle;
    else if (distance <= 25.0f)  return middle_circle;
    else if (distance <= 100.0f) return outer_circle;
    else                         return missed;
}
